bears/general/CPDBear.py: Update language setting

pareksha opened the pull request:
This updates the language settings resulting in prompting
the user if language setting is not provided.
Bear now displays a warning message if it has no support
for the asked language.

Closes https://github.com/coala/coala-bears/issues/1635

<!--
Thanks for your contribution!

Please take a quick look at those things down there. They're quite important.
Really! We wrote them for you. Yes you! With utmost care. Read them.
-->

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

pareksha commented:
@ishanSrt, I think it was a bug, as earlier the bear was not prompting the user to supply the language settings (language setting is mandatory for CPDBear) and was instead throwing error.

ishanSrt commented:
@pareksha if you think this is a bug, get a maintainer to add the label `type/bug` on the issue

ishanSrt commented:
@pareksha FYI it was throwing an error but **was** reminding  that the language parameter is not given
```[ERROR][22:39:50] During execution, we found that some required settings were not provided. They are:
language (from CPDBear) - One of the supported languages of this bear.
[ERROR][22:39:50] An unknown error occurred. This is a bug. We are sorry for the inconvenience. Please contact the developers for assistance. During execution of coala an exception was raised. This should never happen. When asked for, the following information may help investigating:
[INFO][22:39:50] Exception was:
Traceback (most recent call last):

  File "/coala/coalib/coala_main.py", line 81, in run_coala
    arg_list=arg_list)

  File "/coala/coalib/settings/ConfigurationGathering.py", line 387, in gather_configuration
    log_printer)

  File "/coala/coalib/settings/SectionFilling.py", line 42, in fill_settings
    fill_section(section, acquire_settings, log_printer, all_bears)

  File "/coala/coalib/settings/SectionFilling.py", line 88, in fill_section
    new_vals = acquire_settings(log_printer, needed_settings, section)

  File "/coala/coalib/output/Interactions.py", line 26, in fail_acquire_settings
    raise AssertionError

AssertionError
```

pareksha commented:
@ishanSrt, it is clearly specified that ``this is a bug`` in the error message.

ishanSrt commented:
@pareksha Try to understand the difference. `This is a bug` was mentioned by coala as after it found out one of the params were missing while this issue may not be a bug. 
>@pareksha if you think this is a bug, get a maintainer to add the label type/bug on the issue

ishanSrt commented:
also you need to adjust the test cases according to the changes you made

khanchi97 commented on bears/general/CPDBear.py:
```diff
@@ -1,11 +1,9 @@
 from shutil import which
 from xml.etree import ElementTree
-
```
don't remove this blank line.As this maintain a gap between absolute import and relative import I guess.

ishanSrt commented:
Also you need to squash your commits into one

ishanSrt commented:
start by 
```
git reset --soft HEAD~5 &&  
git commit --edit -m"$(git log --format=%B --reverse HEAD..HEAD@{1})"
```
which will open a text editor with concatenation of your last 5 commit messages. Edit that as to be your final commit message and then force push

pareksha commented:
@ishanSrt, I think I know understand the difference between a bug and an issue of this type where missing param was causing an error. I have modified the test cases according to the changes (I was wondering why CircleCI was failing). Also, now there is only one commit. I have removed the irrelevant commits. 
If an incorrect language setting is given, then only the bear will cause a warning.
Make sure to use python with a capital 'P'. This was how the developers have made the language_dict before (didn't feel it right to modify that).
Thanks for reviewing and stating the necessary changes. Please ack if you are satisfied.

pareksha commented:
@khanchi97, the blank line is there again. Thanks for the info. Please acknowledge if satisfied.

pareksha commented:
@newbazz the necessary changes have been made. Please acknowledge if satisfied.

pareksha commented:
@newbazz , what do you mean ? Should I send a new PR ?

newbazz commented:
Nope,rebuild means empty force push ;P

newbazz commented on bears/general/CPDBear.py:
```diff
@@ -73,12 +72,12 @@ def run(self, language: language, minimum_tokens: int=20,
         else:
-            self.err('This bear does not support files with the extension '
-                     "'{}'.".format(language))
+            self.warn('This bear has no support for ' +
```
Why did you changed it to warning?

pareksha replied:
Logging level error is generally used when something went wrong with the code. 
Warning indicates that code is working as expected, something went wrong due to wrong user input.
Hope this clears the doubt @newbazz

ishanSrt commented:
@pareksha you still haven't responded to many of the code reviews I suggested or explanations that were asked. Clarify them first.

ishanSrt commented:
you haven'e even responded to this https://github.com/coala/coala-bears/pull/2205#pullrequestreview-86066585

ishanSrt commented:
nor this https://github.com/coala/coala-bears/pull/2205#pullrequestreview-86067216 @pareksha 
unack https://github.com/coala/coala-bears/pull/2205/commits/065f22f54b0a7fbb932656a7f617adf0868c1907

pareksha commented:
@ishanSrt, I previously told you that the language_dict is made as before, I haven't altered it because this issue is not regarding incomplete language support for CPDBear. It will show a warning if it doesn't have support for the language asked by the user.
Also, CPDBear is working fine now. I will definitely create an issue if any other bear has any such problem. Till now, bears using language settings such as the AnnotationBear are working perfectly fine.
Hope this clears the doubt.

pareksha commented:
@newbazz please acknowledge if you are statisfied.

ishanSrt commented:
@pareksha how does this explain https://github.com/coala/coala-bears/pull/2205#pullrequestreview-86066585 . And you still haven't responded to many code reviews. Respond on them individually. For eg. reason for ditching `coalib.settings.Settin.language`. Please read the reviews thoroughly before responding. Nobody pointed out that there was a problem with `language_dict`

pareksha commented:
@ishanSrt, please read the last reply carefully, I have replied to all your queries. And all the code reviews as well. I am again telling you that it is appropriate that if the language (whether it be blank or some other language) is not recognized by the bear i.e it doesn't have support for that. Then only it will print a warning. Previously, it was not even asking for the language and was thus, showing an error.
Bear is working fine now without ``coalib.settings.Settings.language``. 
Hope this clears all the doubt.

ishanSrt commented:
>Isn't it appropriate that if the language (whether it be blank or some other language) is not recognized by the bear i.e it doesn't have support for that, then it should print a warning?

so you are setting it to blank, so the bear is left with support for no languages? Sorry but I fail to understand this. And even if according to you it does no change, shouldn't you be avoiding unnecessary changes and unnecessary change of variable names?

Also you still haven't responded to https://github.com/coala/coala-bears/pull/2205#pullrequestreview-86066585 what will be the warning or error message if you don't specify any language?

pareksha commented:
@ishanSrt, the language_dict contains the languages which the bear has support for.
The problem with the bear was because of the last variable name. When I changed the variable name and changed the annotation to str, the bear is working fine. Please try to understand.

pareksha commented:
Only necessary changes have been made. :smile:

ishanSrt commented:
@pareksha **ok so now you have started explaining things**
So the reason for changing the variable name is the package `language` conflicting with the var name `language`
I don't think that creates any problem. Check this:
```
>>> import sys
>>> sys=3
>>> print(sys)
3
```
Still got no explanation why you changed it to str. 
>the bear is working fine. Please try to understand.

That way of dealing with code is rather unacceptable. **I think the bear working fine is just a coincident unless you properly explain why you made the changes.**

Still got no answer on 
>Also you still haven't responded to #2205 (review) what will be the warning or error message if you don't specify any language?

*I can't test it out myself because I can't make PMD work on my machine*

jayvdb commented:
@adhikasp , can you have a look at this wrt the changing of `language` to `programming_language`, as it relates to your aspectisation of `language`. Thx

jayvdb commented:
The setting absolutely must be called `language`, and the type annotation must be `language`.

See https://github.com/coala/coala-bears/issues/1635#issuecomment-381801176 for more background.

This PR is way off course.  To begin with, the tests should be replicating the problem reported, which was a language setting of `py`.
